### ISSUE TITLE
Message body color with regexes

### ISSUE THREAD

**Issue by FilippoBonazziSUSE** [neomutt/neomutt#3527]
## Expected Behaviour

I'm trying to color parts of the message body according to regular expressions.

I have this couple of `color` instructions:
```
# Color text formatting
color body underline color229 color234 '_[^_]*_'

# Color some part of markdown
color body color108 color234 '`[^\`]*?`'
color body color108 color234 '^`{3}.*$'
```
which I would expect to do the following:

- color and underline text surrounded by underscores
- color text surrounded by backticks
- color a line such as "```python"

## Actual Behaviour

![20220921_20h41m17s_grim](https://user-images.githubusercontent.com/101266837/191586451-ac2faa5e-93c0-454c-ae53-1e0bbe3a2601.png)

The regex matching seems not to "consume" the underscore characters, highlighting `_*this*_`, `_ and *_` and `_this_` rather than `_*this*_` and `_this_`.

Similarly it highlights everything between any backticks, rather than just the two words.
It also doesn't pick up the "```python" line which should be covered by the third regex.

This is very strange. Am I doing something wrong? I would appreciate any help to investigate the bug or to fix my configuration.

## How often does this happen?

- Always

## NeoMutt Version

```
NeoMutt 20220429

System: Linux 5.19.8-1-default (x86_64)
ncurses: ncurses 6.3.20220910 (compiled with 6.3.20220813)
libidn: 1.38 (compiled with 1.38)
GPGME: 1.16.0-unknown
GnuTLS: 3.7.7
libnotmuch: 5.6.0
PCRE2: 10.40 2022-04-14
storage: kyotocabinet, lmdb
compression: zlib

Configure options: --prefix=/usr --docdir=/usr/share/doc/packages/neomutt --with-mailpath=/var/mail --kyotocabinet --lua --lmdb --gnutls --gpgme --notmuch --sasl --gss --idn --mixmaster --pcre2 --zlib

Compilation CFLAGS: -O2 -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=return-type -flto=auto -g -std=c99 -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D__EXTENSIONS__ -I/usr/include -I/usr/include/lua5.4 -DNCURSES_WIDECHAR

Default options:
  +attach_headers_color +compose_to_sender +compress +cond_date +debug 
  +encrypt_to_self +forgotten_attachments +forwref +ifdef +imap +index_color 
  +initials +limit_current_thread +multiple_fcc +nested_if +new_mail +nntp +pop 
  +progress +quasi_delete +regcomp +reply_with_xorig +sensible_browser +sidebar 
  +skip_quoted +smtp +status_color +timeout +tls_sni +trash 

Compile options:
  -autocrypt +fcntl -flock -fmemopen +futimens +getaddrinfo +gnutls +gpgme +gss 
  +hcache -homespool +idn +inotify -locales_hack +lua +mixmaster +nls +notmuch 
  -openssl +pcre2 +pgp +sasl +smime -sqlite +sun_attachment 

MAILPATH="/var/mail"
MIXMASTER="mixmaster"
PKGDATADIR="/usr/share/neomutt"
SENDMAIL="/usr/sbin/sendmail"
SYSCONFDIR="/etc"
```

## Extra Info

* openSUSE Tumbleweed
* Single neomutt instance
* Running directly in alacritty, no 'screen' or 'tmux'
* Remote IMAP folder (irrelevant)

**Comment by gahr** [comment]
Your second regexp contains a `*?`, which gives me a "repetition-operator operand invalid" error in NeoMutt. If I change it into a `*` I get this

<img width="463" alt="Screen Shot 2022-09-22 at 08 40 49" src="https://user-images.githubusercontent.com/959474/191676270-b0b4b3f3-0f51-46d7-8847-5cf397a15e22.png">

**Comment by FilippoBonazziSUSE** [comment]
Hi, thanks for the prompt reply.

I added the `?` at some point hoping that it would do something, but didn't observe any change. How are you seeing this error? I can't find it in the log even at `-d5`.

Anyway, even by removing the `?` I don't get the highlighting you get. I guess you're running the latest source build? Let me try and see if I can fix the issue by building from the latest source.

**Comment by gahr** [comment]
>  How are you seeing this error? I can't find it in the log even at -d5.

True - you only get an error if you issue the `:color` command interactively from NeoMutt. I'll fix that separately.

> I guess you're running the latest source build?

Also true :) please let me know how it goes with the main branch.

**Comment by gahr** [comment]
#3528 is the fix for reporting the regex compile error at parse time is.

**Comment by FilippoBonazziSUSE** [comment]
Ok, upon further analysis the problem stemmed from another regex I had added a few lines below, and which seemed completely unrelated (a classic):
 ```
color body standout color108 color234 "@?(Filippo)?(Bonazzi)?"
```
For some obscure reason, this does not produce any error upon parsing, and results in the behaviour in the first screenshot.
This is particularly odd since the regex seems unrelated (although I realized that it actually can match the empty string), and also the coloring should be something entirely different (`standout`):
![20220922_18h27m54s_grim](https://user-images.githubusercontent.com/101266837/191801769-4535929d-b4e1-4fa2-be0c-acf245fa83c0.png)

So now I'm just puzzled how the existence of a (admittedly bad) `color body standout A B R1` instruction can modify the behaviour of a `color body X Y R2` a few lines above.

Also none of this depends on neomutt version (I tested again with 20220429).

**Comment by FilippoBonazziSUSE** [comment]
Updated `color` instructions resulting in the screenshot above:

```
# Color text formatting
color body underline color229 color234 '\<_[^_]+_\>'

# Color some part of markdown
color body color108 color234 '`[^\`]+`'
color body color108 color234 '`{3}[^\`]+`{3}'
color body color108 color234 '^`{3}.*$'

# Highlight my name
color body standout color108 color234 "@?(Filippo)?(Bonazzi)?"
```

Of course I realize the last regex is just bad and I will probably just replace it with
```
color body standout color108 color234 "@?Filippo"
```
which works just fine. However it is still unexpected how the instructions affect each other.

**Comment by gahr** [comment]
Yup I can see how the addition of `color body color108 color234 "@?"` messes up the highlighting. I don't know why.
